dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-etcd/src/main/java/org/apache/dolphinscheduler/plugin/registry/etcd/EtcdConnectionStateListener.java

```diff
@@ -0,0 +1,119 @@
+ */
+public class EtcdConnectionStateListener implements AutoCloseable {
+    private final List<ConnectionListener> connectionListeners = Collections.synchronizedList(new ArrayList<>());
+    // A thread pool that periodically obtains connection status
+    private final ScheduledExecutorService scheduledExecutorService;
+    // Client's Channel
+    private AtomicReference<ManagedChannel> channel;
+    // monitored client
+    private Client client;
+    // The state of the last monitor
+    private ConnectionState connectionState;
+    private long initialDelay = 500L;
+    private long delay = 500L;
+    public EtcdConnectionStateListener(Client client) {
+        this.client = client;
+        channel = new AtomicReference<>();
+        this.scheduledExecutorService = Executors.newScheduledThreadPool(
+                1,
+                new ThreadFactoryBuilder().setNameFormat("EtcdConnectionStateListenerThread").setDaemon(true).build());
+    }
+
+    public void addConnectionListener(ConnectionListener connectionListener) {
+        connectionListeners.add(connectionListener);
+    }
+
+    @Override
+    public void close() throws Exception {
+        connectionListeners.clear();
+        scheduledExecutorService.shutdownNow();
+    }
+
+    /**
+     * Apply for a lease through the client, if there is no exception, the connection is normal
+     * @return the current connection state
+     * @throws if there is a exception, return is DISCONNECTED
+     */
+    private ConnectionState isConnected() {
+        try {
+            // Use Get() to ensure Future completes
+            client.getLeaseClient().grant(1).get().getID();
+            return ConnectionState.CONNECTED;
+        } catch (Exception e) {
+            return ConnectionState.DISCONNECTED;
+        }
+    }
+
+    /**
+     * Periodically execute thread to get connection status
+     */
+    public void start() {
+        this.scheduledExecutorService.scheduleWithFixedDelay(() -> {
+            ConnectionState currentConnectionState = isConnected();
+            if (currentConnectionState == connectionState) {
+                return;
+            }
+            if (connectionState == ConnectionState.CONNECTED) {
+                if (currentConnectionState == ConnectionState.DISCONNECTED) {
+                    connectionState = ConnectionState.DISCONNECTED;
+                    triggerListener(ConnectionState.DISCONNECTED);
+                }
+            } else if (connectionState == ConnectionState.DISCONNECTED) {
+                if (currentConnectionState == ConnectionState.CONNECTED) {
+                    connectionState = ConnectionState.CONNECTED;
+                    triggerListener(ConnectionState.RECONNECTED);
+                }
+            } else if (connectionState == null) {
+                connectionState = currentConnectionState;
+                triggerListener(connectionState);
+            }
```

Comment:
Why can't we just simplify this to

```suggestion
            if (connectionState != ConnectionState.SUSPENDED) {
                connectionState = currentConnectionState;
                triggerListener(connectionState);
```

@ruanwenjun I think the api should also notify even if the previous state is suspended and if the listener isn't interested in this they can just ignored.

Reply:
Yes, in fact, for all state changes, we need to notify the listener.
I wrote these logic in MysqlRegistry, since we need to generate `ConnectionState.RECONNECTED` state :), but the state we story in memory is just `RECONNECTED` and `RECONNECTED`.